File: revolut/AccountManager.m
//
//  AccountManager.m
//  revolut
//
//  Created by Alexander Danilov on 02/09/2017.
//  Copyright © 2017 Home. All rights reserved.
//

#import "AccountManager.h"
#import "Account.h"

@implementation AccountManager {
    CurrencyManager *_currencyManager;
}

@synthesize accounts;
    
-(instancetype)initWith:(CurrencyManager *)currencyManager {
    self = [super init];
    if (self) {
        _currencyManager = currencyManager;
        
        accounts = @{
                     @"EUR": [[Account alloc] initWithCurrency:@"EUR" andBalance:[[NSDecimalNumber alloc] initWithInt: 100]],
                     @"GBP": [[Account alloc] initWithCurrency:@"GBP" andBalance:[[NSDecimalNumber alloc] initWithInt: 100]],
                     @"USD": [[Account alloc] initWithCurrency:@"USD" andBalance:[[NSDecimalNumber alloc] initWithInt: 100]]
                     };
    }
    return self;
}

-(BOOL)canExchangeAmount:(NSDecimalNumber *)amount from:(NSString *)fromCurrency {
    Account *fromAccount = accounts[fromCurrency];
    BOOL result = [fromAccount.balance compare:amount] != NSOrderedAscending;
    NSLog(@"canExchange: %@ from: %@ accountAmount: %@ result: %@", amount, fromCurrency, fromAccount.balance, [NSNumber numberWithBool:result]);
    return result;
}

-(BOOL)exchangeAmount:(NSDecimalNumber *)amount from:(NSString *)fromCurrency to:(NSString *)toCurrency error:(NSError **)errorPtr {
    NSLog(@"exchangeAmount: %@ from: %@ to: %@", amount, fromCurrency, toCurrency);
    
    Account *fromAccount = accounts[fromCurrency];
    Account *toAccount = accounts[toCurrency];
    
    BOOL noAccounts = !fromAccount || !toAccount;
    if (noAccounts || ![self canExchangeAmount:amount from:fromCurrency]) {
        NSString *errorMessage = noAccounts ? @"Unknown error" : @"You don't have enough money";
        if (errorPtr) {
            *errorPtr = [NSError errorWithDomain:@"revolut" code:0 userInfo:@{ NSLocalizedDescriptionKey: errorMessage }];
        }
        return FALSE;
    }
    
    [self setValue:[fromAccount.balance decimalNumberBySubtracting:amount]
        forKeyPath:[NSString stringWithFormat:@"accounts.%@.balance", fromCurrency]];
    
    NSDecimalNumber *amountToAdd = [_currencyManager getExchangeRateOfValue:amount from:fromCurrency to:toCurrency];
    [self setValue:[toAccount.balance decimalNumberByAdding: amountToAdd]
        forKeyPath:[NSString stringWithFormat:@"accounts.%@.balance", toCurrency]];
    
    NSLog(@"exchangeAmount Success, newFromAmount: %@ newToAmount: %@", fromAccount.balance, toAccount.balance);
    
    return TRUE;
}

@end

File: revolut/AccountManager.swift
//
//  AccountManager.swift
//  revolut
//
//  Created by Alexander Danilov on 13/08/2018.
//  Copyright © 2018 Home. All rights reserved.
//

import Foundation

class AccountManager: NSObject {
    let currencyManager: CurrencyManager
    
    var accounts: NSDictionary
    
    init(currencyManager: CurrencyManager) {
        self.currencyManager = currencyManager
        self.accounts = NSDictionary(dictionary: [
            "EUR": Account(currency: "EUR", balance: NSDecimalNumber(value: 100)),
            "GBP": Account(currency: "GBP", balance: NSDecimalNumber(value: 100)),
            "USD": Account(currency: "USD", balance: NSDecimalNumber(value: 100))
        ])
        
        super.init()
    }

    func canExchange(amount: NSDecimalNumber, from fromCurrency: String) -> Bool {
        let fromAccount: Account = self.accounts[fromCurrency]! as! Account
        let result: Bool = fromAccount.balance.compare(amount) != .orderedAscending
        print("canExchange: \(amount) from: \(fromCurrency) accountAmount: \(fromAccount.balance) result: \(result)")
        return result
    }
    
    func exchange(amount: NSDecimalNumber, from fromCurrency: String, to toCurrency: String) throws {
        print("exchangeAmount: \(amount) from: \(fromCurrency) to: \(toCurrency)")
        
        let fromAccount: Account = accounts[fromCurrency]! as! Account
        let toAccount: Account = accounts[toCurrency]! as! Account
        if !canExchange(amount: amount, from: fromCurrency) {
            throw NSError(domain: "revolut", code: 0, userInfo: [NSLocalizedDescriptionKey: "You don't have enough money"])
        }
        
        self.setValue(fromAccount.balance.subtracting(amount), forKeyPath: "accounts.\(fromCurrency).balance")
        
        let amountToAdd: NSDecimalNumber = currencyManager.getExchangeRate(ofValue: amount, from: fromCurrency, to: toCurrency)!
        self.setValue(toAccount.balance.adding(amountToAdd), forKeyPath: "accounts.\(toCurrency).balance")

        print("exchangeAmount Success, newFromAmount: \(fromAccount.balance) newToAmount: \(toAccount.balance)")
    }
}
